ActsAsArModel supports more report and rbac use cases

kbrock opened the pull request:
~~BLOCKED ON: #10273~~

Screens that display `ActsAsArModel` are broken. There are a few steps to merge:

- [x] #10264 remove references to `apply_legacy_finder_options`
- [x] #10175 remove `apply_legacy_finder_options` (calling on query blew things up)
- [x] #10273 Fix includes for reports
- This PR (1 commit) - add a test to ensure `ActsAsArModel` works for basic cases

Fixes #10087

Details:
-----

1. Support `includes(:table => {})` (an similar references / order). It is not smart enought to properly the second level of hashes. but `:includes` is not really supported by our `ActsAsArModel` implementations.
2. Support `references(:table => {})`. Though we do not use `:references` in any of our implementations, it was blowing up before. it was free with the `includes()` implementation
3. Support `order(Hash)`. It was free. We may need to support other syntaxes for this as well.
4. Support `length` attribute, which is essentially the same as `size`
5. Support `where(nil)` which is getting called more often now that we are sql munging less.

/cc @lucasponce

Fryguy commented on lib/rbac/filterer.rb:
```diff
@@ -451,6 +448,11 @@ def method_with_scope(ar_scope, options)
     end
 
     def apply_scope(klass, scope)
+      klass = to_class(klass)
```
This ends up calling to_class twice (once on line 177) and then again here.  Is that expected?

kbrock replied:
`to_class` takes a string and calls constantize on it. For a class it is a no-op.

I want just this call, but wasn't able to delete line 177 yet.

It will be a no-op.

kbrock replied:
I'll remove this call to make more clear

Fryguy commented on lib/rbac/filterer.rb:
```diff
@@ -175,9 +175,6 @@ def search(options = {})
-        # could just call all on everything, but that will display deprecation warnings
-        targets = targets.all if targets < ActiveRecord::Base || targets.kind_of?(ActsAsArModel)
-
         scope = apply_scope(targets, scope)
```
I think the naming of targets is throwing me off, because when I think targets I'm thinking an Array (or scope) of objects.  But the to_class call has turned that into a class, so perhaps `targets_class` is a better name for clarity?  Not necessary for this PR, just trying to understand the code because it's confusing.

kbrock replied:
Yea this is confusing. `targets` represents the `:targets` passed in.

Both `:targets` and `:class` are basically the same thing. They can come in as a string/class name, ActiveRecord class, AAArM class, scope, array of ids, or array of objects.

The purpose of this logic is to convert all of those options into a class and scope / `klass` and `scope`

The original code was: `targets = to_class(targets).all` but we were getting deprecation messages.


My long term goal is to merge target and class, options, and reduce the sheer number of valid options.

kbrock replied:
ooh, `to_class` says if it is a string, convert into a class.
And then call `all` on it to convert into a scope.

So the goal of the code is to get it into what you expect as quickly as possible

kbrock commented:
@Fryguy changed the naming a bit. added some comments and added a test class here too to show what is happening.

I tried to update the changes your suggested.

If you note, the blocked PRs are at the beginning of the commit list, but this PR is only 1 commit. and should be pretty close to the original PR that you reviewed.

kbrock commented:
rebased with fixes to other PR. (you can see green dots in commit lists

kbrock commented:
@fryguy unblocked. not sure if you have any further comments

lucasponce commented:
Hi, is there any ETA when all the PRs related to ActsAsArModel changes will be merged ?
After they are in we should validate if we need to perform additional changes on the provider side.
Any info would be welcome :).
/cc @kbrock @abonas
